Fix: Allow RenderPipeline VertexFormat.DrawMode to take effect in GuiRenderer

cputnam-a11y opened the pull request:
Allows pipelines to use the `FabricRenderPipeline$Builder#withUsePipelineDrawModeForGui` to specify whether the pipeline's `VertexFormat.DrawMode` should be respected in `GuiRenderer` this pr does not change vanilla behavior, only allows mods to customize the behavior for their pipelines. An included example renders a triangle to the hud using `VertexFormat.DrawMode.TRIANGES`

Commit:
Fix Gui Rendering only supporting Quads

Commit:
Fix Checkstyle

Commit:
Fix Mixin Names

maityyy commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/impl/client/rendering/FabricRenderPipelineInternals.java:
```diff
@@ -0,0 +1,58 @@
+import net.minecraft.util.Unit;
+
+public final class FabricRenderPipelineInternals {
+	private static final Map<RenderPipeline.Snippet, Optional<Boolean>> extraData = Collections.synchronizedMap(new IdentityHashMap<>());
```
`Optional<Boolean>` can be replaced with `Boolean` and null-check

Commit:
Use null to represent no value for snippets

Commit:
Append extra data to record methods in `RenderPipeline.Snippet`, add test for appended methods

Commit:
Use Map Absence for no value

maityyy commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/mixin/client/rendering/RenderPipelineSnippetMixin.java:
```diff
@@ -30,4 +33,39 @@ class RenderPipelineSnippetMixin implements FabricRenderPipeline.Snippet {
 	}
+
+	@ModifyReturnValue(
+			method = "toString",
```
missing `remap = false`?

modmuss50 replied:
Shouldnt be needed anymore, they are tottaly ingored now: https://github.com/FabricMC/tiny-remapper/commit/0bbcbe4a6ad40eb99c0e6a60f4f21890da4817c2

If someone wants an easy cleanup PR freel free to remove them all.

cputnam-a11y replied:
okay

Commit:
remap = false

Commit:
Revert "remap = false"

This reverts commit 9ba4bc64e0ded50cb165afcc8e10ae71569c4d35.

PepperCode1 commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/mixin/client/rendering/RenderPipelineMixin.java:
```diff
@@ -0,0 +1,40 @@
+@Mixin(RenderPipeline.class)
+public class RenderPipelineMixin implements FabricRenderPipeline {
+	@Unique
+	boolean usePipelineDrawModeForGui = false;
```
This should be private.

cputnam-a11y replied:
done

PepperCode1 commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/mixin/client/rendering/RenderPipelineMixin.java:
```diff
@@ -0,0 +1,40 @@
+	}
+
+	static {
+		FabricRenderPipelineInternals.usePipelineDrawModeForGuiSetter = (pipeline, value) ->
```
While this design works, I am not aware of any other places where Fabric API uses this. I would switch this to a setter method in an internal duck interface that this mixin also implements.

cputnam-a11y replied:
is a :duck:

PepperCode1 commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/impl/client/rendering/FabricRenderPipelineInternals.java:
```diff
@@ -0,0 +1,61 @@
+import net.minecraft.util.Unit;
+
+public final class FabricRenderPipelineInternals {
+	private static final Map<RenderPipeline.Snippet, Boolean> extraData = Collections.synchronizedMap(new IdentityHashMap<>());
```
As discussed on Discord this map should be removed and the boolean should be stored in a mixined field in `Snippet`.

cputnam-a11y replied:
This wasn't the conclusion I saw on discord, and I would rather not do this. If it's needed to get the pr through I'll do it, but the discussion on discord seemed to be closer to "it already happens in fapi"

PepperCode1 replied:
What are the reasons to not make this change?

PepperCode1 replied:
I did say the following.

> I suppose making it weak is good enough

However, I think using a map makes the implementation more complex (for example, it necessitates `MIXIN_PIPELINE_MODIFICATION_REC_GUARD`), and I see no benefit to using a map over a field.

cputnam-a11y replied:
I'll just make it a field and I can cry about it later. there is no reason to block this pr over it, and making the map weak comes with other complications

cputnam-a11y replied:
it's a field now

Commit:
Address Pepper's Review
 - make RenderPipelineMixin.usePipelineDrawModeForGui private
 - replace FabricRenderPipelineInternals.usePipelineDrawModeForGuiSetter with a duck
 - use field over map for storing additional data in RenderPipeline.Snippet :sob:

Commit:
Revert Changes to VanillaBlockModelPartEncoder.java

modmuss50 commented:
Just to confirm this change is only for 1.21.9?

Also looks like you have some checkstyle issues.

MattiDragon commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/mixin/client/rendering/RenderPipelineSnippetMixin.java:
```diff
@@ -0,0 +1,79 @@
+import net.fabricmc.fabric.impl.client.rendering.FabricRenderPipelineInternals;
+
+@Mixin(RenderPipeline.Snippet.class)
+@Debug(export = true)
```
This probably shouldn't be left here

Earthcomputer replied:
Maybe we should deny this in the checkstyle

modmuss50 replied:
Good catch! Might be a good candiate for a checkstyle rule to disallow it.

cputnam-a11y commented:
I think it could be backported all the way to 1.21.6 if needed, but I'll confirm a range in a bit

cputnam-a11y commented:
it should / could be backported all the way to 1.21.6.

Commit:
Fix CheckStyle and Remove `@Debug(export=true)` from `RenderPipelineSnippetMixin`

PepperCode1 commented:
I think 1.21.6/1.21.7 are superseded by 1.21.8, but there could be use in backporting to 1.21.8.

Juuxel commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/api/client/rendering/v1/FabricRenderPipeline.java:
```diff
@@ -0,0 +1,98 @@
+import com.mojang.blaze3d.pipeline.RenderPipeline;
+
+/**
+ * General purpose fabric extensions to the {@link RenderPipeline} class
```
```suggestion
 * General purpose Fabric extensions to the {@link RenderPipeline} class.
```

Juuxel commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/api/client/rendering/v1/FabricRenderPipeline.java:
```diff
@@ -0,0 +1,98 @@
+	}
+
+	/**
+	 * General purpose fabric extensions to the {@link RenderPipeline.Builder} class
```
```suggestion
	 * General purpose Fabric extensions to the {@link RenderPipeline.Builder} class.
```

Juuxel commented on fabric-rendering-v1/src/client/java/net/fabricmc/fabric/api/client/rendering/v1/FabricRenderPipeline.java:
```diff
@@ -0,0 +1,98 @@
+	}
+
+	/**
+	 * General purpose fabric extensions to the {@link RenderPipeline.Snippet} class
```
```suggestion
	 * General purpose Fabric extensions to the {@link RenderPipeline.Snippet} class.
```

cputnam-a11y committed:
Update fabric-rendering-v1/src/client/java/net/fabricmc/fabric/api/client/rendering/v1/FabricRenderPipeline.java

Capitalize Fabric in javadoc

Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

cputnam-a11y committed:
Update fabric-rendering-v1/src/client/java/net/fabricmc/fabric/api/client/rendering/v1/FabricRenderPipeline.java

Capitalize Fabric in javadoc

Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

cputnam-a11y committed:
Update fabric-rendering-v1/src/client/java/net/fabricmc/fabric/api/client/rendering/v1/FabricRenderPipeline.java

Capitalize Fabric in javadoc

Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

modmuss50 commented:
This doesnt merge nicely in 1.21.8, if you want it there please make a seperate PR so we can backport it later.
